publisher-sdk/src/main/java/com/criteo/publisher/BidResponse.java

```diff
@@ -55,8 +64,29 @@ public boolean isBidSuccess() {
     return valid;
   }
 
-  public AdUnitType getAdUnitType() {
-    return adUnitType;
+  @Nullable
+  public String consumeDisplayUrlFor(@NonNull AdUnitType adUnitType) {
+    if (!adUnitType.equals(this.adUnitType)) {
+      return null;
+    }
+
+    return consume(CdbResponseSlot::getDisplayUrl);
+  }
+
+  @Nullable
+  public NativeAssets consumeNativeAssets() {
+    return consume(CdbResponseSlot::getNativeAssets);
+  }
+
+  @Nullable
+  private synchronized <T> T consume(Function1<CdbResponseSlot, T> action) {
+    if (slot == null || slot.isExpired(clock)) {
+      return null;
+    }
+
+    T element = action.invoke(slot);
+    slot = null;
```

Comment:
I think we need some code-documentatin to explain why we are nulling out the slot.

Reply:
Sure, done

Reply:
Is it clear from the publisher point of view that getting a display url consumes the bid? `consumeDisplayUrl` does not indicate that the bid is consumed, but that the displayUrl is consumed. Also, why do we want to have a bid that is usable only once? Is it to workaround the possibility of the publisher caching the bid? But in this case, the publisher can still keep the bid in cache the first time and use it later. I also thought we agreed that it is the backend responsibility to detect when a displayUrl for a given bid is used twice.

@vguerci what are your thoughts on this?

Reply:
Only the `getPrice` method is visible from publishers. Other are internal methods. Let me kotlinify this class, so I can then use `internal` keywords.

The "usable once" logic was already there on InHouse integration. This PR does not add or remove any functionality. At the base, yes I guess it is a way to mitigate the risk that a publisher could save the bid and reuse it multiple time.
Also, even if we can rely on TLA to handle such duplication, I think it is still worth it to skip trivial errors from publisher at client-side level. This is a way to reduce the infra cost.

Reply:
I finally used the `@Internal` annotation to explicit that consume* methods are not for publishers.